### PR TITLE
add alt-text dashboard

### PR DESCRIPTION
**Pull request by dnoneill**
part of https://github.com/sul-dlss/exhibits/issues/2628

Screenshot 2024-11-14 at 7 59 52 PM

Screenshot 2024-11-14 at 7 59 25 PM

Small updates done so it now has icons and sorts differently than above but everything else looks the same.

Screenshot 2024-11-21 at 2 20 12 PM

### CODE REVIEW

File: app/controllers/spotlight/accessibility_controller.rb [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+      @total_alt_items = @pages.sum { |page| page[:can_have_alt_text] }
+
+      attach_dashboard_breadcrumbs
+    end
+
+    def get_alt_info(page)
+      can_have_alt_text = 0
+      has_alt_text = 0
+      page.content.each do |content|
+        content.item.each_value do |item|
```
Comment (corylown) [inline app/controllers/spotlight/accessibility_controller.rb:30]: I think you'll need account for cases where there's a widget on the page that doesn't have items? When I add a widget like the iframe widget or the header widget to a page I get an error here:

```
NoMethodError in Spotlight::AccessibilityController#alt_text
undefined method `each_value' for nil:NilClass

/spotlight/app/controllers/spotlight/accessibility_controller.rb:31:in `block in get_alt_info'
/spotlight/app/controllers/spotlight/accessibility_controller.rb:29:in `each'
/spotlight/app/controllers/spotlight/accessibility_controller.rb:29:in `get_alt_info'
/spotlight/app/controllers/spotlight/accessibility_controller.rb:19:in `block in alt_text' 
```

Comment (dnoneill) [reply app/controllers/spotlight/accessibility_controller.rb]: Yep.

**Comment by alundgard** [comment]
Sidebar and header organization
spotlight-sidebar-a11y

**Comment by alundgard** [comment]
Maybe change `Of the images displayed via spotlight's widgets...` to `Of the images displayed via widgets...` so we don't need to tweak this text for Exhibits.

File: app/javascript/spotlight/admin/blocks/resources_block.js [hunk 2]
```diff
@@ -8,18 +8,16 @@ Core.Block.Resources = (function(){
     formable: true,
     autocompleteable: true,
     show_heading: true,
-    show_alt_text: true,
```
Comment (dnoneill) [inline app/javascript/spotlight/admin/blocks/resources_block.js:11]: I tried to mirror description and not make this a separate function. Every time I did I would get an error.

File: app/models/sir_trevor_rails/block.rb [hunk 3]
```diff
@@ -18,12 +18,20 @@ def format
       send(:[], :format).present? ? send(:[], :format).to_sym : DEFAULT_FORMAT
     end
 
+    def self.alt_text?
```
Comment (corylown) [inline app/models/sir_trevor_rails/block.rb:21]: Is there a reason this is defined as a class method and not an instance method? You're calling it here https://github.com/projectblacklight/spotlight/pull/3264/files#diff-50de7f3fa74c59817c7b61c1adf10178b1039ba6b7c95529962359aef2a4f13fR18 as if it were an instance method.

File: app/models/sir_trevor_rails/blocks/solr_documents_block.rb [hunk 4]
```diff
@@ -13,6 +13,10 @@ def with_solr_helper(solr_helper)
         @solr_helper = solr_helper
       end
 
+      def self.alt_text?
```
Comment (corylown) [inline app/models/sir_trevor_rails/blocks/solr_documents_block.rb:16]: Should this be an instance method?

File: app/models/sir_trevor_rails/blocks/solr_documents_embed_block.rb [hunk 5]
```diff
@@ -5,6 +5,9 @@ module Blocks
     ##
     # Embed documents (using a special blacklight view configuration) and text block
     class SolrDocumentsEmbedBlock < SirTrevorRails::Blocks::SolrDocumentsBlock
+      def self.alt_text?
```
Comment (corylown) [inline app/models/sir_trevor_rails/blocks/solr_documents_embed_block.rb:8]: Should this be an instance method?

File: app/models/sir_trevor_rails/blocks/uploaded_items_block.rb [hunk 6]
```diff
@@ -12,6 +12,10 @@ def files
         (item || {}).map { |_, file| file }.select { |file| file[:display].to_s == 'true' }
       end
 
+      def self.alt_text?
```
Comment (corylown) [inline app/models/sir_trevor_rails/blocks/uploaded_items_block.rb:15]: Should this be an instance method?

File: app/controllers/spotlight/accessibility_controller.rb [hunk 7]
```diff
@@ -0,0 +1,49 @@
+# frozen_string_literal: true
+
+module Spotlight
+  ##
+  # Exhibit dashboard controller
+  class AccessibilityController < Spotlight::ApplicationController
+    before_action :authenticate_user!
+    load_and_authorize_resource :exhibit, class: Spotlight::Exhibit
+    helper_method :get_alt_info
```
Comment (corylown) [inline app/controllers/spotlight/accessibility_controller.rb:9]: I don't see where this method is being called as a helper.

File: app/controllers/spotlight/accessibility_controller.rb [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+      { can_have_alt_text:, has_alt_text:, page: }
+    end
+
+    protected
+
+    def attach_dashboard_breadcrumbs
+      add_breadcrumb(t(:'spotlight.exhibits.breadcrumb', title: @exhibit.title), @exhibit)
+      add_breadcrumb(t(:'spotlight.accessibility.header'), exhibit_dashboard_path(@exhibit))
+      add_breadcrumb(t(:'spotlight.accessibility.alt_text.header'), exhibit_alt_text_path(@exhibit))
+    end
```
Comment (corylown) [inline app/controllers/spotlight/accessibility_controller.rb:47]: Can these both be private methods? I don't see where else they're used outside of this class.

**Review by corylown — CHANGES_REQUESTED** [review]

File: app/models/sir_trevor_rails/block.rb (hunk 3 above)
Comment (dnoneill) [reply app/models/sir_trevor_rails/block.rb:21]: Yeah these lines need it to be a class method: https://github.com/projectblacklight/spotlight/pull/3264/files#diff-9da975442affa1f40c188c0b08c324ac361b8b3e51d484e78d5633de76dc92e1R31-R34

File: app/models/sir_trevor_rails/blocks/solr_documents_block.rb (hunk 4 above)
Comment (dnoneill) [reply app/models/sir_trevor_rails/blocks/solr_documents_block.rb:16]: See above

File: app/models/sir_trevor_rails/blocks/solr_documents_embed_block.rb (hunk 5 above)
Comment (dnoneill) [reply app/models/sir_trevor_rails/blocks/solr_documents_embed_block.rb:8]: See above

File: app/controllers/spotlight/accessibility_controller.rb (hunk 8 above)
Comment (dnoneill) [reply app/controllers/spotlight/accessibility_controller.rb]: Yep. That is very doable.

File: app/models/sir_trevor_rails/block.rb (hunk 3 above)
Comment (dnoneill) [reply app/models/sir_trevor_rails/block.rb:25]: Which I now realize you need both. It was an instance until I added the line above. I have updated the model to call the classes' variable for the instance https://github.com/projectblacklight/spotlight/pull/3264/files#diff-9da975442affa1f40c188c0b08c324ac361b8b3e51d484e78d5633de76dc92e1R21-R23

**Review by corylown — APPROVED** [review]

**Pull request merged**